gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -213,16 +262,62 @@ public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedE
     return isTerminated();
   }
 
+  /** Scheduling loop. */
+  private ScheduledFuture<?> scheduleNextRefresh() {
+    long delayPeriod = REFRESH_PERIOD.toMillis();
+    long jitter = (long) ((Math.random() - 0.5) * JITTER_PERCENTAGE * delayPeriod);
+    long delay = jitter + delayPeriod;
+    return channelRefreshExecutorService.schedule(
+        () -> {
+          try {
+            refresh();
+          } finally {
+            scheduleNextRefresh();
+          }
+        },
+        delay,
+        TimeUnit.MILLISECONDS);
+  }
+
   /**
-   * Performs a simple round robin on the list of {@link ManagedChannel}s in the {@code channels}
-   * list.
+   * Replace all of the channels in the channel pool with fresh ones. This is meant to mitigate the
+   * hourly GFE disconnects by giving clients the ability to prime the channel on reconnect.
    *
-   * @return A {@link ManagedChannel} that can be used for a single RPC call.
+   * <p>This is done on a best effort basis. If the replacement channel fails to construct, the old
+   * channel will continue to be used.
    */
-  private ManagedChannel getNextChannel() {
-    return getChannel(indexTicker.getAndIncrement());
+  @InternalApi("Visible for testing")
+  void refresh() {
+    List<Entry> localEntries = entries.get();
+    ArrayList<Entry> newEntries = new ArrayList<>(localEntries);
+    ArrayList<Entry> removedEntries = new ArrayList<>();
+
+    for (int i = 0; i < localEntries.size(); i++) {
+      try {
+        Entry removed = newEntries.set(i, new Entry(channelFactory.createSingleChannel()));
+        removedEntries.add(removed);
+      } catch (IOException e) {
+        LOG.log(Level.WARNING, "Failed to refresh channel, leaving old channel", e);
+      }
+    }
+    entries.set(ImmutableList.copyOf(newEntries));
+
+    removedEntries.forEach(Entry::requestShutdown);
   }
 
+  /**
+   * Get and retain a Channel Entry. The returned will have its rpc count incremented, preventing it
+   * from getting recycled.
+   */
+  Entry getRetainedEntry(int affinity) {
+    for (int i = 0; i < 5; i++) {
```

Comment:
How about adding a comment like "the maximum number of concurrent calls to this method for any given time span is at most 2, so the loop can actually be 2 times. But going for 5 times for a safety margin for potential code evolving"?

Reply:
done